internal/session/session.go

```diff
@@ -234,23 +186,121 @@ func (s *Session) GetBlock(parent context.Context, k cid.Cid) (blocks.Block, err
 // returns a channel that found blocks will be returned on. No order is
 // guaranteed on the returned blocks.
 func (s *Session) GetBlocks(ctx context.Context, keys []cid.Cid) (<-chan blocks.Block, error) {
-	ctx = logging.ContextWithLoggable(ctx, s.uuid)
+	// If there are no keys supplied, just return a closed channel
+	out := make(chan blocks.Block)
+	if len(keys) == 0 {
+		close(out)
+		return out, nil
+	}
 
-	return bsgetter.AsyncGetBlocks(ctx, s.ctx, keys, s.notif,
-		func(ctx context.Context, keys []cid.Cid) {
-			select {
-			case s.incoming <- op{op: opWant, keys: keys}:
-			case <-ctx.Done():
-			case <-s.ctx.Done():
+	// Use a WantRequest to listen for incoming messages pertaining to the keys
+	wr, err := s.wrm.NewWantRequest(keys)
+	if err != nil {
+		return nil, err
+	}
+
+	// Listens for incoming messages
+	listen := func() {
+		// When the function exits
+		defer func() {
+			// Close the channel of outgoing blocks
+			close(out)
+
+			// Mark any remaining wants as no longer wanted
+			remaining := wr.OnClosing()
+
+			// Inform the session of the cancelled keys
+			s.blockingEnqueue(op{
+				op:   opCancel,
+				keys: remaining,
+			})
+
+			// Cancel the remaining wants and call Close() on the WantRequest
+			// when all processing of the wants has completed. Note that we
+			// need to make sure that cancels are sent after any corresponding
+			// wants that may still be in the pipeline.
+			s.sws.Cancel(remaining, wr.Close)
```

Comment:
This breaks the abstraction. We're adding keys by sending an "op" to the session but canceling by calling this function?